client-server-client/src/main/scala/zio/raft/client/RaftClient.scala

```diff
@@ -0,0 +1,615 @@
+                _ <- ZIO.logWarning(s"Connection timeout, trying next address: $nextAddr")
+                _ <- transport.disconnect().orDie
+                _ <- transport.connect(nextAddr).orDie
+                newNonce <- Nonce.generate()
+                _ <- transport.sendMessage(ContinueSession(sessionId, newNonce)).orDie
+                now <- Clock.instant
+              } yield copy(nonce = newNonce, currentAddressIndex = nextIndex, createdAt = now)
+            } else {
+              ZIO.succeed(this)
+            }
+          } yield nextState
+        
+        case StreamEvent.ServerMsg(other) =>
+          ZIO.logWarning(s"Unexpected message in ConnectingExistingSession: ${other.getClass.getSimpleName}").as(this)
+        
+        case StreamEvent.Action(ClientAction.Disconnect) =>
+          for {
+            _ <- transport.disconnect().orDie
+          } yield Disconnected
+        
+        case StreamEvent.KeepAliveTick =>
+          ZIO.succeed(this)
+      }
+    }
+  }
+  
+  case class Connected(
+    sessionId: SessionId,
+    capabilities: Map[String, String],
+    createdAt: Instant,
+    serverRequestTracker: ServerRequestTracker,
+    nextRequestId: Ref[RequestId],
+    pendingRequests: PendingRequests,
+    addresses: List[String],
+    currentAddressIndex: Int
+  ) extends ClientState {
+    override def stateName: String = s"Connected($sessionId)"
+    
+    override def handle(
+      event: StreamEvent,
+      transport: ZmqClientTransport,
+      config: ClientConfig,
+      serverRequestQueue: Queue[ServerRequest]
+    ): UIO[ClientState] = {
+      event match {
+        case StreamEvent.Action(ClientAction.Disconnect) =>
+          for {
+            _ <- transport.sendMessage(CloseSession(ClientShutdown)).orDie
+            _ <- transport.disconnect().orDie
+            _ <- ZIO.logInfo("Disconnected")
+          } yield Disconnected
+        
+        case StreamEvent.Action(ClientAction.SubmitCommand(payload, promise)) =>
+          for {
+            requestId <- nextRequestId.updateAndGet(_.next)
+            now <- Clock.instant
+            request = ClientRequest(requestId, payload, now)
+            _ <- transport.sendMessage(request).orDie
+            newPending = pendingRequests.add(requestId, payload, promise, now)
+          } yield copy(pendingRequests = newPending)
+        
+        case StreamEvent.ServerMsg(ClientResponse(requestId, result)) =>
+          pendingRequests.complete(requestId, result) match {
+            case Some(newPending) =>
+              ZIO.succeed(copy(pendingRequests = newPending))
+            case None =>
+              ZIO.logWarning(s"Response for unknown request: $requestId").as(this)
+          }
+        
+        case StreamEvent.ServerMsg(KeepAliveResponse(_)) =>
+          ZIO.succeed(this)
+        
+        case StreamEvent.ServerMsg(serverRequest: ServerRequest) =>
+          if (serverRequestTracker.shouldProcess(serverRequest.requestId)) {
```

Comment:
I don't like this piece of code... serverRequestTracker.shouldProcess should return an enum of 3 options, `process`, `oldReuqest` and `outOfOrder`. It would be more elegant.